appscale/tools/appscale_stats.py

```diff
@@ -0,0 +1,657 @@
+  """
+  login_host = LocalState.get_login_host(keyname=keyname)
+  login_acc = AppControllerClient(
+    host=login_host,
+    secret=LocalState.get_secret_key(keyname)
+  )
+  all_private_ips = login_acc.get_all_private_ips()
+  cluster_stats = login_acc.get_cluster_stats()
+  roles_data = {
+    ip:
+      next(
+        n["roles"] for n in cluster_stats if n["private_ip"] == ip
+      )
+    for ip in all_private_ips
+  }
+
+  return roles_data
+
+
+def get_node_stats(raw_node_stats, all_roles, specified_roles, verbose):
+  """
+  Obtaines useful information from node statistics and returns:
+  PRIVATE IP, AVAILABLE MEMORY, LOADAVG, PARTITIONS USAGE, ROLES values.
+
+  Args:
+    raw_node_stats: A dist in which each key is an ip and value is a dict
+      of useful information.
+    all_roles: A dist in which each key is an ip and value is a role list.
+    specified_roles: A list representing specified roles
+      that nodes should contain.
+    verbose: A boolean - add all partitions if True,
+      add only three the most used partitions if False.
+
+  Returns:
+    A list of node statistics headers.
+    A list of node statistics.
+  """
+  node_stats_headers = [
+    "PRIVATE IP", "AVAILABLE MEMORY", "LOADAVG", "PARTITIONS USAGE", "ROLES"
+  ]
+
+  node_stats = []
+  for ip, node in raw_node_stats.iteritems():
+    ip_roles = all_roles[ip]
+
+    if specified_roles:
+      matches = [role for role in specified_roles if role in ip_roles]
+      if not matches:
+        continue
+
+    if "shadow" in ip_roles:
+      node_info = [
+        get_marked(ip, "bold"),
+        get_marked(render_memory(memory=node["memory"]), "bold"),
+        get_marked(render_loadavg(loadavg=node["loadavg"]), "bold"),
+        get_marked(
+          render_partitions(partitions=node["partitions_dict"],
+                            verbose=verbose), "bold"
+        ),
+        get_marked(u", ".join(ip_roles), "bold")
+      ]
+    else:
+      node_info = [
+        ip,
+        render_memory(memory=node["memory"]),
+        render_loadavg(loadavg=node["loadavg"]),
+        render_partitions(partitions=node["partitions_dict"], verbose=verbose),
+        u", ".join(ip_roles)
+      ]
+    node_stats.append(node_info)
+
+  node_stats.sort(key=lambda n: n[0], reverse=False)
+
+  return node_stats_headers, node_stats
+
+
+def get_process_stats(raw_process_stats):
+  """
+  Obtains useful information from process statistics and returns:
+  PRIVATE IP, MONIT NAME, UNIQUE MEMORY (MB), CPU (%) values.
+
+  Args:
+    raw_process_stats: A dist in which each key is an ip and value is a dict
+      of useful information.
+
+  Returns:
+    A list of process statistics headers.
+    A list of process statistics.
+  """
+  process_stats_headers = [
+    "PRIVATE IP", "MONIT NAME", "UNIQUE MEMORY (MB)", "CPU (%)"
+  ]
+
+  process_stats = []
+  for ip, node in raw_process_stats.iteritems():
+    stat = node.get("processes_stats")
```

Comment:
`get` usually indicates that the function should proceed if the key doesn't exist. You could either do `node.get('processes_stats', [])` (since an iterable is expected below) or `node['processes_stats']` if an exception should be thrown.